src/main/java/net/mcreator/workspace/references/ReferencesFinder.java

```diff
@@ -46,134 +46,111 @@
-			if (anyValueMatches(ge, String.class, e -> e.isAnnotationPresent(ModElementReference.class), (a, t) -> {
-				ModElementReference ref = a.getAnnotation(ModElementReference.class);
-				return !Set.of(ref.defaultValues()).contains(t) && query.equals(
-						GeneratorWrapper.getElementPlainName(t));
-			})) {
-				elements.add(me);
-			} else if (anyValueMatches(ge, MappableElement.class, e -> e.isAnnotationPresent(ModElementReference.class),
-					(a, t) -> query.equals(GeneratorWrapper.getElementPlainName(t.getUnmappedValue())))) {
-				elements.add(me);
-			} else if (anyValueMatches(ge, Procedure.class, e -> e.isAnnotationPresent(ModElementReference.class),
-					(a, t) -> t.getName() != null && !t.getName().isEmpty() && !t.getName().equals("null")
-							&& query.equals(t.getName()))) {
-				elements.add(me);
-			} else if (anyValueMatches(ge, String.class, e -> e.isAnnotationPresent(BlocklyXML.class),
-					(a, t) -> t.contains(">CUSTOM:" + query + "</field>") || t.contains(">" + query + "</field>"))) {
-				elements.add(me);
-			}
-		});
+	/**
+	 * Convenience method to provide a list of all generatable elements contained in the provided workspace in a thread-safe manner.
+	 * List returned by this method can be used in parallel streams as long as all other methods in the said parallel stream are
+	 * also thread safe. Thus, e.g. calling ModElementManager within the parallel stream later would still cause problems.
+	 *
+	 * @param workspace Workspace to obtain generatable elements from.
+	 * @return List of all generatable elements contained in the provided workspace.
+	 */
+	private static List<GeneratableElement> getGeneratableElements(Workspace workspace) {
+		return workspace.getModElements().stream().map(ModElement::getGeneratableElement).collect(Collectors.toList());
+	}
 
-		return elements;
+	public static Set<ModElement> searchModElementUsages(Workspace workspace, ModElement element) {
+		//@formatter:off
+		return getGeneratableElements(workspace).parallelStream()
+			.filter(ge ->
+				anyValueMatches(ge, String.class, e -> e.isAnnotationPresent(ModElementReference.class), (a, t) -> {
+					ModElementReference ref = a.getAnnotation(ModElementReference.class);
+					return !Set.of(ref.defaultValues()).contains(t) && element.getName().equals(
+							GeneratorWrapper.getElementPlainName(t));
+				}) ||
+				anyValueMatches(ge, MappableElement.class, e -> e.isAnnotationPresent(ModElementReference.class), (a, t) ->
+					element.getName().equals(GeneratorWrapper.getElementPlainName(t.getUnmappedValue()))
+				) ||
+				anyValueMatches(ge, Procedure.class, e -> e.isAnnotationPresent(ModElementReference.class), (a, t) ->
+					element.getName().equals(t.getName())
+				) ||
+				anyValueMatches(ge, String.class, e -> e.isAnnotationPresent(BlocklyXML.class), (a, t) ->
+					t.contains(">CUSTOM:" + element.getName() + "</field>") || t.contains(">" + element.getName() + "</field")

```

Comment:
The way I implemented this does not allow searching for ME items (e.g. buckets or spawn eggs) in Blockly XML. Do you think this can be fixed without too many changes?

Reply:
Good catch, exactly the reason why the more people test, the better, I completely missed that when testing :)

I can provide a fix for this

Reply:
This regex should do the trick: https://regexr.com/7lo3k

I have done quite some manual testing but more is welcome of course